src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "io/io_queue.h"
+
+#include "io/logger.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sleep.hh>
+
+namespace experimental::io {
+
+io_queue::io_queue(
+  persistence* storage,
+  std::filesystem::path path,
+  completion_callback_type complete) noexcept
+  : storage_(storage)
+  , path_(std::move(path))
+  , complete_(std::move(complete)) {}
+
+void io_queue::start() noexcept {
+    log.debug("Starting io queue for {}", path_);
+    dispatcher_ = dispatch();
+}
+
+seastar::future<> io_queue::stop() noexcept {
+    log.debug("Stopping io queue for {}", path_);
+    auto dispatcher = std::exchange(
+      dispatcher_, seastar::make_ready_future<>());
+    stop_ = true;
+    cond_.signal();
+    co_await std::move(dispatcher);
+    co_await gate_.close();
+}
+
+const std::filesystem::path& io_queue::path() const noexcept { return path_; }
+
+seastar::future<> io_queue::open() noexcept {
+    log.debug("Opening {}", path_);
+    assert(!opened());
+    assert(running_.empty());
+    file_ = co_await storage_->open(path_);
+    cond_.signal();
+}
+
+bool io_queue::opened() const noexcept { return file_ != nullptr; }
+
+/*
+ * closing the file always succeeds. once seastar::file::close is invoked the
+ * handle is not longer usable, so we swallow any errors here and release units
+ * to avoid blocking other queues as there isn't much else we can do.
+ *
+ * failure to close a file is quite rare outside of the normal cases like trying
+ * to close an unknown or already closed file handle. these cases won't occur
+ * using the seastar api since it prevents those mistakes. in any case, what we
+ * care about most when closing a file is that its file descriptor resource is
+ * freed in the kernel to avoid reaching an open file limit.
+ */
+seastar::future<> io_queue::close() noexcept {
+    log.debug("Closing {}", path_);
+    assert(opened());
+
+    // ensure all on-going ops have completed
+    auto units = co_await seastar::get_units(
+      ops_, seastar::semaphore::max_counter());
+
+    auto file = std::exchange(file_, nullptr);
+    try {
+        co_await file->close();
+    } catch (...) {
+        log.warn(
+          "Unable to close file {}: {}", path_, std::current_exception());
+    }
+}
+
+void io_queue::submit_read(page& page) noexcept {
+    page.set_flag(page::flags::read);
```

Comment:
should we check `is_linked` here? I can enqueue same page twice by mistake and boost intrusive list will trigger assertion

Reply:
If a page has been queued for read, it must not be re-queued for read or write until has completed. I didn't mention this in the comment, so I added a comment to address this.

Keep in mind that the I/O queue is not intended for general use. It is an internal component.

```diff
diff --git a/src/v/io/io_queue.h b/src/v/io/io_queue.h
index 0937a99a5c..026f155b43 100644
--- a/src/v/io/io_queue.h
+++ b/src/v/io/io_queue.h
@@ -107,6 +107,9 @@ public:

     /**
      * Submit a read I/O request.
+     *
+     * A page submitted for read must not be resubmitted for read or write until
+     * it has completed.
      */
     void submit_read(page&) noexcept;
```